### DISCUSSION TITLE
Some errors encountered when performing yaw error simulation

### DISCUSSION THREAD

**Discussion by tatesnow** [OpenFAST/openfast#991] [Q&A] [answered]
I want to analyze the effect of yaw angle on the wind turbine load, so I modified the yaw angle parameter （NacYaw） of ElastoDyn. The yaw angle is taken from 0 to 25 degrees with an interval of 1, and I found that the simulation can run normally when the yaw angle is small, but the simulation fails when the yaw angle is more than 19 degrees.

The error message is shown below：

**************************************************************************************************
OpenFAST

Copyright (C) 2021 National Renewable Energy Laboratory
Copyright (C) 2021 Envision Energy USA LTD

This program is licensed under Apache License Version 2.0 and comes with ABSOLUTELY NO WARRANTY.
See the "LICENSE" file distributed with this software for details.
**************************************************************************************************

OpenFAST-v3.0.0
Compile Info:
 - Compiler: Intel(R) Fortran Compiler 1900
 - Architecture: 64 bit
 - Precision: single
 - OpenMP: No
 - Date: Jun 23 2021
 - Time: 21:25:05
Execution Info:
 - Date: 02/07/2022
 - Time: 11:56:50+0800

OpenFAST input file heading:
    FAST          Certification - Test #18: NREL 5.0 MW Baseline Wind Turbine (Onshore)

Running ElastoDyn.
Nodal outputs section of ElastoDyn input file not found or improperly formatted.
Running AeroDyn.
AD15 Nodal Outputs: Nodal output section of AeroDyn input file not found or improperly formatted.
Skipping nodal outputs.
Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0. (node 1, blade 1)
Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0. (node 2, blade 1)
Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0. (node 3, blade 1)
Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0. (node 4, blade 1)
Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0. (node 1, blade 2)
Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0. (node 2, blade 2)
Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0. (node 3, blade 2)
Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0. (node 4, blade 2)
Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0. (node 1, blade 3)
Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0. (node 2, blade 3)
Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0. (node 3, blade 3)
Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0. (node 4, blade 3)
Running InflowWind.

   Reading a 39x34 grid (190 m wide, 7.5 m to 172.5 m above ground) with a characteristic wind
   speed of 3 m/s. This full-field file was generated by TurbSim (v2.00.07a-bjj, 14-Jun-2016) on
   04-Jan-2022 at 20:16:20.

   Processed 12420 time steps of 20-Hz full-field data (period of 621 seconds).
Running ServoDyn.
Running ServoDyn Interface for Bladed Controllers (using Intel Visual Fortran for Windows, ).
Using legacy Bladed DLL interface.
 Time: 0 of 620 seconds.

FAST_Solution0:CalcOutputs_And_SolveForInputs:SolveOption2:SrvD_CalcOutput:DLL_controller_call:Run
ning with torque and pitch control of the NREL offshore 5MW baseline wind turbine from DISCON.dll
as written by J. Jonkman of NREL/NWTC for use in the IEA Annex XXIII OC3 studies.

Warning: SkewedWakeCorrection encountered a large value of chi (90.407 deg), so the yaw
correction will be limited. This warning will not be repeated though the condition may persist.
See the AD15 chi output channels, and consider turning off the Pitt/Peters skew model (set
SkewMod=1) if this condition persists.

FAST_Solution:CalcOutputs_And_SolveForInputs:SolveOption2:RotCalcOutput:BEMT_CalcOutput(node 5,
blade 1):UA_CalcOutput:UA_BlendSteady:Temporarily turning off UA due to high angle of attack or
low relative velocity. This warning will not be repeated though the condition may persist.

FAST_Solution:CalcOutputs_And_SolveForInputs:SolveOption2:RotCalcOutput:BEMT_CalcOutput(node 18,
blade 2):UA_CalcOutput:Mach number exceeds 0.3. Theory is invalid. This warning will not be
repeated though the condition may persist.

FAST_Solution:FAST_AdvanceStates:AD_UpdateStates:BEMT_UpdateStates:UpdatePhi(node 5, blade
1):BEMT_UnCoupledSolve:There is no valid value of phi for these operating conditions: Vx =
10.678, Vy = -50.143, rlocal = 11.74, theta = 0.23228, geometric phi = 2.9318. This warning will
not be repeated though the condition may persist. (See GeomPhi output channel.)

FAST_Solution:FAST_AdvanceStates:ED_ABM4:ED_CalcContStateDeriv:SetCoordSy:Small angle assumption
violated in SUBROUTINE SmllRotTrans() due to a large blade deflection (ElastoDyn SetCoordSy). The
solution may be inaccurate. Simulation continuing, but future warnings from SmllRotTrans() will
be suppressed.
 Additional debugging message from SUBROUTINE SmllRotTrans(): 3.75000E-02 s

The BEM solution is being turned off due to low TSR.  (TSR = -123.53). This warning will not be
repeated though the condition may persist. (See GeomPhi output channel.)

FAST_Solution:CalcOutputs_And_SolveForInputs:SolveOption2:RotCalcOutput:BEMT_CalcOutput(node 19,
blade 3):UA_CalcOutput:Mach number exceeds 1.0. Equations cannot be evaluated.

 OpenFAST encountered an error at simulation time 0.10625 of 620 seconds.
 Simulation error level: FATAL ERROR

 Aborting OpenFAST.

**Comment by jjonkman** [comment, marked as answer]
Dear @tatesnow,

Do you have the nacelle-yaw DOF enabled in ElastoDyn (`YawDOF` = True)?  If so, have you set the neutral yaw angle in ServoDyn equal to the initial yaw angle in ElastoDyn, i.e., `YawNeut` = `NacYaw`.  If not, the nacelle-yaw angle will quickly snap back to 0deg and result in unrealistic loads and deflections.  If you want to keep a nonzero yaw angle and keep the nacelle-yaw DOF enabled, then set both `YawNeut` and `NacYaw` to the same value.

Best regards,

**Reply by tatesnow** [reply]
Dear @jjonkman ,
Thank you very much! According to your suggestion, I fixed my error!
Best regards,

**Comment by icnosvbovwb** [comment]
Dear @jjonkman 
![image](https://user-images.githubusercontent.com/107483730/205646404-6439814d-1878-4ce3-a1e3-a92f0c01c741.png)
I run the ROSCO-2.6.0 controller with the 32-bit dll compiled by openfast3.1. The example used is IEA-15-240-RWT-UMaineSemi. When I run the turbulent wind, there will be an error in the picture. The whole process only takes more than 900 steps. Do you know where the problem occurs?

**Reply by jjonkman** [reply]
Dear @icnosvbovwb,

It looks like your model has gone unstable.  Presumably you are using the OpenFAST model of the IEA Wind 15-MW reference wind turbine top the UMaine semi available from here: https://github.com/IEAWindTask37/IEA-15-240-RWT/tree/master/OpenFAST/IEA-15-240-RWT-UMaineSemi.  What did you change in this model that results in the issue you report?

Best regards,

**Comment by icnosvbovwb** [comment]
Dear @jjonkman 
Thank you for your reply. I haven't changed the dat file. After running more than 900 steps with rosco, the data starts to be abnormal, but it can still be calculated, but there is no data displayed later, and NAN appears.I will display all the DAT files I use in the following figure：
![image](https://user-images.githubusercontent.com/107483730/205942382-538a34ea-b202-4f76-a9fb-7bb6e93ff1d4.png)
![image](https://user-images.githubusercontent.com/107483730/205936180-b9303e97-4558-4a2a-8165-30fd19311d10.png)
![image](https://user-images.githubusercontent.com/107483730/205936245-58abbe1f-61b8-4a47-aa0b-78d58a8ee719.png)
![image](https://user-images.githubusercontent.com/107483730/205936327-95dc79f7-bdc6-4f62-b2c4-bab5606e42a2.png)
![image](https://user-images.githubusercontent.com/107483730/205936411-40f3d7cb-7c4d-46ee-8c76-482079259812.png)
![image](https://user-images.githubusercontent.com/107483730/205936515-c0f85139-36b3-400b-a1ef-4ffc9e2c4ced.png)
![image](https://user-images.githubusercontent.com/107483730/205936579-4c0a663e-44d3-4113-843c-c49786425a09.png)
![image](https://user-images.githubusercontent.com/107483730/205936656-a457afbe-278a-4444-80d7-983b85979875.png)
![image](https://user-images.githubusercontent.com/107483730/205936726-13499f7d-ca91-409e-b9e4-54ce8c334586.png)
![image](https://user-images.githubusercontent.com/107483730/205937483-c13d94ec-6e17-4a35-ab90-e80f269747c7.png)
![image](https://user-images.githubusercontent.com/107483730/205937587-ad2dbd72-1a69-4dac-b4ed-aa55a9276427.png)

**Reply by jjonkman** [reply]
Dear @icnosvbovwb,

I haven't reviewed all of your files (PNG files are not great for reviewing and comparing), but it looks like you did change some inputs relative to the OpenFAST model available from the IEA Wind Task 37 repository (https://github.com/IEAWindTask37/IEA-15-240-RWT/tree/master/OpenFAST/IEA-15-240-RWT-UMaineSemi), e.g., the nacelle-yaw DOF has been enabled and the initial conditions have been changed in ElastoDyn, the controller in ServoDyn has changed from the ROSCO dynamic library to a Simulink interface, and you've changed the wind speed and wind type.  Can you clarify why you changed these inputs and does one of these changes result in the instability?

I'm not too familiar with the Simulink implementation of ROSCO, which you are presumably using, but it is likely the controller is sensitive to the initial rotor speed and pitch angle.  The initial rotor speed, in particular, does not look correct for the wind speed you are simulating, which could cause incorrect behavior in the controller.

Best regards,
